### PR TITLE
change(consensus): Add lockbox funding stream

### LINKED ISSUE #8686: Add validation for NU6 lockbox pool
## Motivation

All of the current dev fund proposals include a lockbox pool for a reserve.

## Possible Design

Update `check::subsidy_is_valid()` function used by the zebra-consensus block verifier to check that the coinbase transaction's contribution to the lockbox pool is correct.

### LINKED ISSUE #8687: Update coinbase transaction generation in `getblocktemplate` RPC method for NU6
## Motivation

There will likely be changes to the coinbase transaction consensus rules in NU6 and how coinbase transactions should be contructed.

### LINKED ISSUE #8701: change(consensus): Validate post-NU6 subsidies and miner fees

### PR DESCRIPTION
**Pull request by arya2**
## Motivation

We likely want to add a lockbox funding stream.

Will close #8686, and close #8687 once parameters/addresses are updated.

Part of #8695.

This PR will need to be updated once there's more information about which dev fund proposal we should implement, and, if it's alternative 2 in [this draft ZIP](https://zips.z.cash/draft-nuttycom-funding-allocation), also what funding stream addresses to use for ZCG.

### Specifications & References

Alternatives 2 and 3 in https://zips.z.cash/draft-nuttycom-funding-allocation
https://zips.z.cash/draft-nuttycom-lockbox-streams

## Solution

- Adds a new funding stream receiver for the "deferred pool"
- Checks funding streams after the next halving
- Generates coinbase transactions with new funding streams after NU6 is active 
- Checks that block subsidies are fully claimed after NU6 is active
- Checks that coinbase transactions spend less than:
   `block_subsidy + miner_fees - expected_lockbox_funding_stream_amount`
- Adds a `lockbox_input_value()` function 

Related changes:
- Factors `Halving(height)` calculation into its own `num_halvings()` function
- In `block::check::subsidy_is_valid()`:
  - Removes unnecessary check that `halving_div.count_ones() == 1`
  - Removes redundant check for `height > slow_start_shift`
  - Checks `num_halvings()` instead of `halving_divisor()`
  - Minor simplifications
 
### Tests

- Adds an acceptance test that relies on configurable funding streams on Testnet
- Updates test to check NU6 funding stream values
- Adds a `check_lockbox_input_value()` test for the `lockbox_input_value()` fn

### Follow Up Work

- https://github.com/ZcashFoundation/zebra/pull/8729
- https://github.com/ZcashFoundation/zebra/pull/8727

Blocked on ZIP-253 status being changed to 'Proposed':
- Set an NU6 activation height for Testnet 
- [Update minimum network protocol versions for NU6 on Testnet/Mainnet
](https://github.com/ZcashFoundation/zebra/pull/8733)
- Update initial current minimum network protocol version
- Update post-NU6 funding stream height ranges
- Update the NU6 consensus branch id
- [Update the end of support halt time](https://github.com/ZcashFoundation/zebra/pull/8734)

This PR also adds some TODOs for documentation updates that should be addressed promptly.

### PR Author's Checklist

- [x] The PR name will make sense to users.
- [ ] The PR provides a CHANGELOG summary.
- [x] The solution is tested.
- [x] The documentation is up to date.
- [x] The PR has a priority label.

### PR Reviewer's Checklist

- [ ] The PR Author's checklist is complete.
- [ ] The PR resolves the issue.

### CODE REVIEW

File: zebra-consensus/src/block/check.rs [hunk 1]
```diff
@@ -193,8 +193,15 @@ pub fn subsidy_is_valid(block: &Block, network: &Network) -> Result<(), BlockErr
         //
         // https://zips.z.cash/protocol/protocol.pdf#fundingstreams
         for (receiver, expected_amount) in funding_streams {
-            let address =
-                subsidy::funding_streams::funding_stream_address(height, network, receiver);
+            if receiver == FundingStreamReceiver::Deferred {
```
Comment (arya2) [inline zebra-consensus/src/block/check.rs:197]: Consider moving logic from `miner_fees_are_valid()` to `subsidy_is_valid()`, returning the expected total input value of the miner fees, and only checking that the miner fees are more than or exactly the same as expected (depending on the whether NU6 is active and block subsidies must be fully claimed).

File: zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs [hunk 2]
```diff
@@ -58,15 +58,16 @@ pub struct FundingStream {
     ///
     /// The current Zcash funding streams only use transparent addresses,
     /// so Zebra doesn't support Sapling addresses in this RPC.
-    pub address: transparent::Address,
+    #[serde(skip_serializing_if = "Option::is_none")]
+    pub address: Option<transparent::Address>,
```
Comment (upbqdn) [inline zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs]: This changes the format of the RPC response since the field is not originally optional: https://zcash.github.io/rpc/getblocksubsidy.html. I wonder if this could cause some mining pools to crash.

File: zebra-consensus/src/block/subsidy/general.rs [hunk 3]
```diff
@@ -108,6 +111,40 @@ pub fn miner_subsidy(height: Height, network: &Network) -> Result<Amount<NonNega
...
+        .get(&network.kind())
+        .expect("must have funding stream height range on all networks");
+
+    // `min(height, last_height_with_deferred_pool_contribution) - (nu6_activation_height - 1)`,
+    // funding stream height range end bound is not incremented since it's an exclusive end bound
+    let num_blocks_with_lockbox_output = height
+        .next()
+        .expect("should be a valid height")
+        .min(funding_stream_height_range.end)
+        - nu6_activation_height;
```
Comment (upbqdn) [inline zebra-consensus/src/block/subsidy/general.rs]: If the caller provides a height lower than `funding_stream_height_range` and `nu6_activation_height-1`, say, 0, then `num_blocks_with_lockbox_output` will be negative, and the function will panic at its second last `expect`.

File: zebra-consensus/src/block/subsidy/general.rs [hunk 4]
```diff
@@ -108,6 +111,40 @@ pub fn miner_subsidy(height: Height, network: &Network) -> Result<Amount<NonNega
...
+    let &deferred_amount_per_block = funding_stream_values(nu6_activation_height, network)
+        .expect("we always expect a funding stream hashmap response even if empty")
+        .get(&FundingStreamReceiver::Deferred)
+        .expect("we expect a lockbox funding stream after NU5");
+
+    let funding_stream_height_range = FUNDING_STREAM_HEIGHT_RANGES
+        .get(&network.kind())
+        .expect("must have funding stream height range on all networks");
+
+    // `min(height, last_height_with_deferred_pool_contribution) - (nu6_activation_height - 1)`,
```
Comment (upbqdn) [inline zebra-consensus/src/block/subsidy/general.rs:139]:
```suggestion
    // `min(height, last_height_with_deferred_pool_contribution) - (nu6_activation_height - 1)`,
    // We decrement NU6 activation height since it's an inclusive lower bound.
```

File: zebra-consensus/src/block/check.rs [hunk 5]
```diff
@@ -245,9 +257,17 @@ pub fn miner_fees_are_valid(
...
+        .map_err(|_| SubsidyError::SumOverflow)?;
+
+    let should_allow_unclaimed_subsidy =
+        NetworkUpgrade::current(network, height) < NetworkUpgrade::Nu6;
 
-    if left > right {
+    if if should_allow_unclaimed_subsidy {
+        left > right
+    } else {
+        left != right
```
Comment (upbqdn) [inline zebra-consensus/src/block/check.rs]:
```suggestion
        // TODO: Document the need for claiming all subsidy once the ZIP for NU6 is out.
        left != right
```

File: zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs (hunk 2 above)
Comment (upbqdn) [reply zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs]: I opened #8702.

File: zebra-consensus/src/block/check.rs (hunk 1 above)
Comment (conradoplg) [reply zebra-consensus/src/block/check.rs:197]: Also this part only runs before the next halving (`halving_div < 4` above). We'll need to ensure these checks also run after the next halving, either by adding a new `else if` or modifying this one.

File: zebra-consensus/src/block/subsidy/funding_streams.rs [hunk 6]
```diff
@@ -25,14 +28,20 @@ pub fn funding_stream_values(
...
-    if height >= canopy_height {
+    if current_network_upgrade >= Canopy {
         let range = FUNDING_STREAM_HEIGHT_RANGES.get(&network.kind()).unwrap();
         if range.contains(&height) {
             let block_subsidy = block_subsidy(height, network)?;
-            for (&receiver, &numerator) in FUNDING_STREAM_RECEIVER_NUMERATORS.iter() {
+            let funding_stream_numerators = if current_network_upgrade < Nu6 {
+                PRE_NU6_FUNDING_STREAM_RECEIVER_NUMERATORS.iter()
+            } else {
+                POST_NU6_FUNDING_STREAM_RECEIVER_NUMERATORS.iter()
```
Comment (conradoplg) [inline zebra-consensus/src/block/subsidy/funding_streams.rs]: (I added this comment before I saw that the plan was to change `FUNDING_STREAM_HEIGHT_RANGES`, but also see my comment there.)

The new funding streams won't last forever. While we could add a check later it's probably safer to add it now. Both ZIPs being considered specify an end height so we might define a constant like `NU6_FUNDING_STREAM_END_HEIGHT` and check for it, and modify the height when the final ZIP is chosen.

File: zebra-consensus/src/parameters/subsidy.rs [hunk 7]
```diff
@@ -72,29 +75,45 @@ lazy_static! {
...
+    };
+
     /// Start and end Heights for funding streams
     /// as described in [protocol specification §7.10.1][7.10.1].
     ///
     /// [7.10.1]: https://zips.z.cash/protocol/protocol.pdf#zip214fundingstreams
     // TODO: Move the value here to a field on `testnet::Parameters` (#8367)
     pub static ref FUNDING_STREAM_HEIGHT_RANGES: HashMap<NetworkKind, std::ops::Range<Height>> = {
         let mut hash_map = HashMap::new();
+        // TODO: Adjust these values once a proposal is selected
```
Comment (conradoplg) [inline zebra-consensus/src/parameters/subsidy.rs]: The halving already happened on the testnet, so the height range there is discontinuous (it stopped and will start back again). We will probably need another map for the NU6 funding streams.

File: zebra-consensus/src/block/check.rs (hunk 1 above)
Comment (arya2) [reply zebra-consensus/src/block/check.rs:197]: > this part only runs before the next halving

Whoops, thank you! Fixed in https://github.com/ZcashFoundation/zebra/pull/8694/commits/8f00f227556f1229141c20e34284a8c3af4c6213 (along with a minor refactor/simplification, let me know if it would be better to minimize the changes and only update that one condition)

File: zebra-consensus/src/block/subsidy/general.rs (hunk 3 above)
Comment (arya2) [reply zebra-consensus/src/block/subsidy/general.rs]: Nice catch! Fixed in https://github.com/ZcashFoundation/zebra/pull/8694/commits/90893bee9b2089c43134ab2550b9da1d6cd17477.

File: zebra-consensus/src/block/subsidy/general.rs (hunk 4 above)
Comment (arya2) [reply zebra-consensus/src/block/subsidy/general.rs:139]: Applied in https://github.com/ZcashFoundation/zebra/pull/8694/commits/b8d4a7975f32eae2bc125c7da3a7be6ff5bacea9.

File: zebra-consensus/src/block/check.rs (hunk 5 above)
Comment (arya2) [reply zebra-consensus/src/block/check.rs]: Applied in https://github.com/ZcashFoundation/zebra/pull/8694/commits/b8d4a7975f32eae2bc125c7da3a7be6ff5bacea9.

File: zebra-consensus/src/parameters/subsidy.rs (hunk 7 above)
Comment (arya2) [reply zebra-consensus/src/parameters/subsidy.rs]: Yep. Added another map in https://github.com/ZcashFoundation/zebra/pull/8694/commits/90893bee9b2089c43134ab2550b9da1d6cd17477.

File: zebrad/tests/acceptance.rs [hunk 8]
```diff
@@ -3217,3 +3217,118 @@ async fn trusted_chain_sync_handles_forks_correctly() -> Result<()> {
...
     Ok(())
 }
+
+/// Test successful block template submission as a block proposal on a custom Testnet.
+///
+/// This test can be run locally with:
+/// `RUSTFLAGS='--cfg zcash_unstable="nu6"' cargo test --package zebrad --test acceptance --features getblocktemplate-rpcs -- nu6_lockbox_funding_stream --exact --show-output`
+#[tokio::test(flavor = "multi_thread")]
+#[cfg(all(feature = "getblocktemplate-rpcs", zcash_unstable = "nu6"))]
+async fn nu6_lockbox_funding_stream() -> Result<()> {
```
Comment (arya2) [inline zebrad/tests/acceptance.rs]: This doesn't actually test what we want yet, I'll open another PR to add the funding stream height ranges to `testnet::Parameters` so it works as expected.

Update: It should be working now.

File: zebra-chain/Cargo.toml [hunk 9]
```diff
@@ -176,3 +176,6 @@ required-features = ["bench"]
 [[bench]]
 name = "redpallas"
 harness = false
+
+[lints.rust]
```
Comment (upbqdn) [inline zebra-chain/Cargo.toml:180]: ###### Optional Suggestion

```suggestion
# TODO: Remove these once they're no longer needed for NU6.
[lints.rust]
```

File: zebrad/Cargo.toml [hunk 10]
```diff
@@ -296,4 +296,4 @@ zebra-grpc = { path = "../zebra-grpc", version = "0.1.0-alpha.5" }
 zebra-utils = { path = "../zebra-utils", version = "1.0.0-beta.38" }
 
 [lints.rust]
```
Comment (upbqdn) [inline zebrad/Cargo.toml:301]: ###### Optional Suggestion

```suggestion
# TODO: Remove 'cfg(zcash_unstable, values("nu6"))' once it's no longer needed for NU6.
[lints.rust]
```

File: zebra-rpc/src/methods/get_block_template_rpcs/get_block_template.rs [hunk 11]
```diff
@@ -164,6 +164,13 @@ where
     Tip: ChainTip + Clone + Send + Sync + 'static,
     SyncStatus: ChainSyncStatus + Clone + Send + Sync + 'static,
 {
+    // TODO:
+    // - Add a `disable_peers` field to `Network` to check instead of `disable_pow()` (#8361)
+    // - Check the field in `sync_status` so it applies to the mempool as well.
+    if network.disable_pow() {
+        return Ok(());
```
Comment (upbqdn) [inline zebra-rpc/src/methods/get_block_template_rpcs/get_block_template.rs:172]: ###### Optional Suggestion

Let's mention in the function docs that it returns an `Ok` when PoW is disabled.

File: zebra-chain/Cargo.toml (hunk 9 above)
Comment (arya2) [reply zebra-chain/Cargo.toml:180]: Applied in https://github.com/ZcashFoundation/zebra/pull/8694/commits/3a730b7fcaba623179703daa6797cb886e6dc381.

File: zebrad/Cargo.toml (hunk 10 above)
Comment (arya2) [reply zebrad/Cargo.toml:301]: Applied in https://github.com/ZcashFoundation/zebra/pull/8694/commits/3a730b7fcaba623179703daa6797cb886e6dc381.

File: zebra-rpc/src/methods/get_block_template_rpcs/get_block_template.rs (hunk 11 above)
Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs/get_block_template.rs:172]: Added to the function docs in https://github.com/ZcashFoundation/zebra/pull/8694/commits/3a730b7fcaba623179703daa6797cb886e6dc381.

**Comment by oxarbitrage** [comment]
For cross reference here is the zcashd ongoing implementation of the lockbox: https://github.com/zcash/zcash/pull/6912

**Review by conradoplg — CHANGES_REQUESTED** [review]
See comment about a possible panic

**Comment by oxarbitrage** [comment]
The commit from https://github.com/ZcashFoundation/zebra/pull/8723 should be out of the diff of this PR. 

A tedious way to do it is to clone main, cherry pick the commits (fix conflicts if needed) and force push. Can probably be done with rebase but i am not sure.

Let me know if you want me to do that.

File: zebra-chain/src/parameters/network_upgrade.rs [hunk 12]
```diff
@@ -214,8 +216,7 @@ pub(crate) const CONSENSUS_BRANCH_IDS: &[(NetworkUpgrade, ConsensusBranchId)] =
     (Heartwood, ConsensusBranchId(0xf5b9230b)),
     (Canopy, ConsensusBranchId(0xe9ff75a6)),
     (Nu5, ConsensusBranchId(0xc2d6d0b4)),
-    // TODO: Use the real consensus branch ID once it's specified.
-    (Nu6, ConsensusBranchId(0xdeadc0de)),
+    (Nu6, ConsensusBranchId(0xc8e71055)),
```
Comment (arya2) [inline zebra-chain/src/parameters/network_upgrade.rs:219]:
```suggestion
    (Nu6, ConsensusBranchId(0x7fffffff)),
```

File: zebra-chain/src/parameters/network_upgrade.rs [hunk 13]
```diff
@@ -214,8 +216,7 @@ pub(crate) const CONSENSUS_BRANCH_IDS: &[(NetworkUpgrade, ConsensusBranchId)] =
     (Heartwood, ConsensusBranchId(0xf5b9230b)),
     (Canopy, ConsensusBranchId(0xe9ff75a6)),
     (Nu5, ConsensusBranchId(0xc2d6d0b4)),
-    // TODO: Use the real consensus branch ID once it's specified.
-    (Nu6, ConsensusBranchId(0xdeadc0de)),
+    (Nu6, ConsensusBranchId(0x7fffffff)),
```
Comment (arya2) [inline zebra-chain/src/parameters/network_upgrade.rs]: Note: This will have to be the same as the `zcash_primitives` consensus branch id for nu6 for the acceptance test in this PR to pass.

Update: Added a note to the acceptance test docs.

File: zebrad/tests/acceptance.rs [hunk 14]
```diff
@@ -3228,3 +3228,301 @@ async fn trusted_chain_sync_handles_forks_correctly() -> Result<()> {
 
     Ok(())
 }
+
+/// Test successful block template submission as a block proposal or submission on a custom Testnet.
+///
+/// This test can be run locally with:
+/// `RUSTFLAGS='--cfg zcash_unstable="nu6"' cargo test --package zebrad --test acceptance --features getblocktemplate-rpcs -- nu6_lockbox_funding_stream --exact --show-output`
```
Comment (arya2) [inline zebrad/tests/acceptance.rs]:
```suggestion
/// `RUSTFLAGS='--cfg zcash_unstable="nu6"' cargo test --package zebrad --test acceptance --features getblocktemplate-rpcs -- nu6_funding_streams_and_coinbase_balance --exact --show-output`
```

Update: Fixed in https://github.com/ZcashFoundation/zebra/pull/8694/commits/4627c0854dc40d65cf7a33ee16645a7c8834ef48.

File: zebra-consensus/src/block/subsidy/funding_streams.rs [hunk 15]
```diff
@@ -52,8 +52,17 @@ pub fn funding_stream_values(
...
 /// [7.10]: https://zips.z.cash/protocol/protocol.pdf#fundingstreams
-fn funding_stream_address_index(height: Height, network: &Network) -> usize {
+fn funding_stream_address_index(
+    height: Height,
+    network: &Network,
+    receiver: FundingStreamReceiver,
+) -> Option<usize> {
+    if receiver == FundingStreamReceiver::Deferred {
+        return None;
+    }
```
Comment (arya2) [inline zebra-consensus/src/block/subsidy/funding_streams.rs:62]: @conradoplg This and the other changes in https://github.com/ZcashFoundation/zebra/pull/8694/commits/7df2e2863f75c2b85323e8906f9f30ee0bacc73f should fix the panic issue you noticed.

File: zebra-consensus/src/block/check.rs [hunk 16]
```diff
@@ -246,9 +260,19 @@ pub fn miner_fees_are_valid(
     // https://zips.z.cash/protocol/protocol.pdf#txnconsensus
     let left = (transparent_value_balance - sapling_value_balance - orchard_value_balance)
         .map_err(|_| SubsidyError::SumOverflow)?;
-    let right = (block_subsidy + block_miner_fees).map_err(|_| SubsidyError::SumOverflow)?;
+    let right = (block_subsidy + block_miner_fees - expected_deferred_amount)
+        .map_err(|_| SubsidyError::SumOverflow)?;
+
+    // TODO: Add link to exact coinbase balance ZIP
+    let should_allow_unclaimed_subsidy =
+        NetworkUpgrade::current(network, height) <= NetworkUpgrade::Nu5;
```
Comment (daira) [inline zebra-consensus/src/block/check.rs]: Although this is probably fine in practice, I note that it *assumes* that whatever upgrade is next after NU5 will activate the ZIP for https://zips.z.cash/draft-hopwood-coinbase-balance . I suggest gating `should_allow_unclaimed_subsidy` specifically on NU6 not being activated.

File: zebra-consensus/src/block/check.rs [hunk 17]
```diff
@@ -246,9 +260,19 @@ pub fn miner_fees_are_valid(
...
+    let is_invalid_miner_fee = if should_allow_unclaimed_subsidy {
+        left > right
+    } else {
+        left != right
+    };
 
-    if left > right {
+    if is_invalid_miner_fee {
         Err(SubsidyError::InvalidMinerFees)?;
     }
```
Comment (daira) [inline zebra-consensus/src/block/check.rs]: The current title ("change(consensus): Add lockbox funding stream") of the PR does not reflect that it includes this change. Also, I would suggest making it a separate PR for ease of review. (The other changes aren't dependent on it, I think.)

File: zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs (hunk 2 above)
Comment (daira) [reply zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs]: zcashd will use [a separate "lockboxstreams" subobject](https://github.com/zcash/zcash/pull/6912/files#diff-ccc24453c13307f815879738d3bf00eec351417537fbf10dde1468180cacd2f1R938-R945) for that reason and as agreed on #8702.

File: zebra-consensus/src/block/check.rs (hunk 17 above)
Comment (arya2) [reply zebra-consensus/src/block/check.rs]: Reverted changes related to the coinbase-should-balance-exactly ZIP in https://github.com/ZcashFoundation/zebra/pull/8694/commits/4627c0854dc40d65cf7a33ee16645a7c8834ef48, opened https://github.com/ZcashFoundation/zebra/pull/8727 instead.

File: zebra-consensus/src/block/check.rs (hunk 16 above)
Comment (arya2) [reply zebra-consensus/src/block/check.rs]: > I suggest gating should_allow_unclaimed_subsidy specifically on NU6 not being activated.

Applied in https://github.com/ZcashFoundation/zebra/pull/8727.

File: zebrad/tests/acceptance.rs [hunk 18]
```diff
@@ -3228,3 +3228,257 @@ async fn trusted_chain_sync_handles_forks_correctly() -> Result<()> {
...
 }
+
+/// Test successful block template submission as a block proposal or submission on a custom Testnet.
+///
+/// This test can be run locally with:
+/// `RUSTFLAGS='--cfg zcash_unstable="nu6"' cargo test --package zebrad --test acceptance --features getblocktemplate-rpcs -- nu6_funding_streams_and_coinbase_balance --exact --show-output`
+/// Note: The consensus branch id for NU6 in Zebra (defined in the `CONSENSUS_BRANCH_IDS` constant) must match the one in the `zcash_protocol` crate, currently `0xc8e71055`.
+#[tokio::test(flavor = "multi_thread")]
+#[cfg(all(feature = "getblocktemplate-rpcs", zcash_unstable = "nu6"))]
+async fn nu6_funding_streams_and_coinbase_balance() -> Result<()> {
```
Comment (arya2) [inline zebrad/tests/acceptance.rs:3239]:
```suggestion
async fn nu6_funding_streams() -> Result<()> {
```

Comment (arya2) [inline zebrad/tests/acceptance.rs]:
```suggestion
/// `RUSTFLAGS='--cfg zcash_unstable="nu6"' cargo test --package zebrad --test acceptance --features getblocktemplate-rpcs -- nu6_funding_streams --exact --show-output`
/// Note: The consensus branch id for NU6 in Zebra (defined in the `CONSENSUS_BRANCH_IDS` constant) must match the one in the `zcash_protocol` crate, currently `0xc8e71055`.
#[tokio::test(flavor = "multi_thread")]
#[cfg(all(feature = "getblocktemplate-rpcs", zcash_unstable = "nu6"))]
async fn nu6_funding_streams() -> Result<()> {
```

Update: Applied in https://github.com/ZcashFoundation/zebra/pull/8694/commits/126fdab77fce5b4ecfa8e7265b4fea795889478c.

File: zebra-chain/src/parameters/network/subsidy.rs [hunk 19]
```diff
@@ -74,6 +77,8 @@ impl FundingStreamReceiver {
             FundingStreamReceiver::Ecc => "Electric Coin Company",
             FundingStreamReceiver::ZcashFoundation => "Zcash Foundation",
             FundingStreamReceiver::MajorGrants => "Major Grants",
+            // TODO: Find out what this should be called and update the funding stream name
+            FundingStreamReceiver::Deferred => "Deferred Fund",
```
Comment (PaulLaux) [inline zebra-chain/src/parameters/network/subsidy.rs]: consider updating documentation for this function.

File: zebra-chain/src/parameters/network/tests/vectors.rs [hunk 20]
```diff
@@ -398,7 +402,7 @@ fn check_configured_funding_stream_constraints() {
             recipients: Some(vec![ConfiguredFundingStreamRecipient {
                 receiver: FundingStreamReceiver::Ecc,
                 numerator: 10,
-                addresses: vec![],
+                addresses: Some(vec![]),
```
Comment (PaulLaux) [inline zebra-chain/src/parameters/network/tests/vectors.rs:405]: Since there are some logical similarity between `None` and an empty vector, I would avoid introducing `Option<>` here and stick with a unified handling.

File: zebra-chain/src/parameters/network_upgrade.rs [hunk 21]
```diff
@@ -124,7 +125,8 @@ pub(super) const TESTNET_ACTIVATION_HEIGHTS: &[(block::Height, NetworkUpgrade)]
     (block::Height(903_800), Heartwood),
     (block::Height(1_028_500), Canopy),
     (block::Height(1_842_420), Nu5),
-    // TODO: Add NU6.
+    // TODO: Add NU6
+    // (block::Height(2_942_000), Nu6),
```
Comment (PaulLaux) [inline zebra-chain/src/parameters/network_upgrade.rs:129]: think you can already add this under the `nu6` flag. (x2)

File: zebra-chain/src/parameters/network/subsidy.rs (hunk 19 above)
Comment (oxarbitrage) [inline zebra-chain/src/parameters/network/subsidy.rs]: This is maybe going to be called "Lockbox": https://github.com/zcash/zcash/pull/6912/files#diff-e3c8b938334507afb1fb2cd91583f9cb8ce5fad39dc8b4c54db2205d34634afaR39-R44

We can do it in a separated PR after zcashd confirms.

File: zebra-chain/src/parameters/network/subsidy.rs [hunk 22]
```diff
@@ -229,8 +251,21 @@ lazy_static! {
     // TODO: Add a reference to lockbox stream ZIP, this is currently based on the number of blocks between the
     //       start and end heights for Mainnet in https://zips.z.cash/draft-nuttycom-funding-allocation
     pub static ref POST_NU6_FUNDING_STREAMS_TESTNET: FundingStreams = FundingStreams {
+        // TODO: Adjust this height range and recipient list once a proposal is selected
         height_range: Height(2_942_000)..Height(3_362_000),
```
Comment (oxarbitrage) [inline zebra-chain/src/parameters/network/subsidy.rs]: I think we need to add this numbers as constants so we can have a more clear description on what they are and how/when they will be changed.

File: zebra-consensus/src/block/subsidy/general.rs [hunk 23]
```diff
@@ -118,10 +118,44 @@ pub fn output_amounts(transaction: &Transaction) -> HashSet<Amount<NonNegative>>
         .collect()
 }
 
+/// Lockbox funding stream total input value for a block height.
+///
+/// Assumes a constant funding stream amount per block.
+// TODO: Cache the lockbox value balance in zebra-state (will be required for tracking lockbox
+//       value balance after ZSF ZIPs or after a ZIP for spending from the deferred pool)
```
Comment (oxarbitrage) [inline zebra-consensus/src/block/subsidy/general.rs]: What `ZSF ZIPs` means ?

File: zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs (hunk 2 above)
Comment (oxarbitrage) [reply zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs]: I think the changes for `getblocksubsidy` and `getblocktemplate` RPCs should be reversed if possible and addressed in the context of https://github.com/ZcashFoundation/zebra/issues/8702.

File: zebrad/tests/acceptance.rs [hunk 24]
```diff
@@ -3228,3 +3228,257 @@ async fn trusted_chain_sync_handles_forks_correctly() -> Result<()> {
 
     Ok(())
 }
+
+/// Test successful block template submission as a block proposal or submission on a custom Testnet.
```
Comment (oxarbitrage) [inline zebrad/tests/acceptance.rs:3232]: Can we move this test to a different PR ?

**Review by oxarbitrage — COMMENTED** [review]
Left some comments and questions.

File: zebra-consensus/src/block/check.rs [hunk 25]
```diff
@@ -246,7 +260,8 @@ pub fn miner_fees_are_valid(
     // https://zips.z.cash/protocol/protocol.pdf#txnconsensus
     let left = (transparent_value_balance - sapling_value_balance - orchard_value_balance)
         .map_err(|_| SubsidyError::SumOverflow)?;
-    let right = (block_subsidy + block_miner_fees).map_err(|_| SubsidyError::SumOverflow)?;
+    let right = (block_subsidy + block_miner_fees - expected_deferred_amount)
+        .map_err(|_| SubsidyError::SumOverflow)?;
```
Comment (PaulLaux) [inline zebra-consensus/src/block/check.rs:270]: The comment above `left = ..` should be updated.

**Review by PaulLaux — COMMENTED** [review]
added some minor comments

File: zebra-chain/src/parameters/network_upgrade.rs (hunk 21 above)
Comment (arya2) [reply zebra-chain/src/parameters/network_upgrade.rs:129]: We can (and I originally did), but we want to wait at least until the NU6 deployment ZIP status is "Proposed" before setting an any activation heights, and perhaps until the release PRs for the versions of Zebra that deploy NU6 on Mainnet/Testnet.

File: zebra-chain/src/parameters/network/subsidy.rs (hunk 19 above)
Comment (arya2) [reply zebra-chain/src/parameters/network/subsidy.rs]: I updated in https://github.com/ZcashFoundation/zebra/pull/8694/commits/bf4943270366430d218fb6a4900fc5ab76aba7b3, we can update it again if it changes.

File: zebra-chain/src/parameters/network/subsidy.rs (hunk 22 above)
Comment (arya2) [reply zebra-chain/src/parameters/network/subsidy.rs]: Moved to constants in https://github.com/ZcashFoundation/zebra/pull/8694/commits/bf4943270366430d218fb6a4900fc5ab76aba7b3.

File: zebra-consensus/src/block/subsidy/general.rs (hunk 23 above)
Comment (arya2) [reply zebra-consensus/src/block/subsidy/general.rs]: Ah, sorry, it was referring to the Zcash Sustainability Fund ZIPs, updated in https://github.com/ZcashFoundation/zebra/pull/8694/commits/bf4943270366430d218fb6a4900fc5ab76aba7b3.

It's specifically [ZIP 234](https://github.com/zcash/zips/pull/706) that will cause an issue, we can probably remove this function altogether in https://github.com/ZcashFoundation/zebra/pull/8729.

File: zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs (hunk 2 above)
Comment (arya2) [reply zebra-rpc/src/methods/get_block_template_rpcs/types/subsidy.rs]: > I think the changes for getblocksubsidy and getblocktemplate RPCs should be reversed if possible and addressed in the context of https://github.com/ZcashFoundation/zebra/issues/8702.

We need the changes to the `getblocktemplate` RPC method for it to produce valid NU6 block templates, and we need to change `getblocksubsidy` in this PR since it changes `funding_stream_address()` to return an `Option`. 

Updated `getblocksubsidy` to exclude the lockbox funding stream from its response in https://github.com/ZcashFoundation/zebra/pull/8694/commits/14fbfb454f9fd58cb4c96b8ca9821048bb699a4f since we agreed we want to add information about the lockbox funding stream(s) in a different list.

File: zebrad/tests/acceptance.rs (hunk 24 above)
Comment (arya2) [reply zebrad/tests/acceptance.rs:3232]: I'd like to keep it here if it doesn't need significant changes.

File: zebra-chain/src/parameters/network/subsidy.rs (hunk 19 above)
Comment (arya2) [reply zebra-chain/src/parameters/network/subsidy.rs]: Added a TODO for referencing the NU6 funding stream ZIP once that's been assigned a ZIP number in https://github.com/ZcashFoundation/zebra/pull/8694/commits/7987c7636548d54831d50f1c8215b9ad8c553fe1.

File: zebra-chain/src/parameters/network/tests/vectors.rs (hunk 20 above)
Comment (arya2) [reply zebra-chain/src/parameters/network/tests/vectors.rs:405]: Good point. I think I wanted it for deserializing config files that omit the field, but mistakenly updated this field type too. Updated in https://github.com/ZcashFoundation/zebra/pull/8694/commits/e5eaf38ba10e56a51f02234d80d548a164e0033f.

File: zebra-consensus/src/block/check.rs (hunk 25 above)
Comment (arya2) [reply zebra-consensus/src/block/check.rs:270]: Updated the comment in https://github.com/ZcashFoundation/zebra/pull/8694/commits/e5eaf38ba10e56a51f02234d80d548a164e0033f.

@daira this likely needs an update to the protocol specification if it's correct.

**Review by oxarbitrage — APPROVED** [review]

File: zebra-chain/src/parameters/network/tests/vectors.rs (hunk 20 above)
Comment (conradoplg) [reply zebra-chain/src/parameters/network/tests/vectors.rs:405]: I think it's still an Option in the ConfiguredFundingStreamRecipient? Was your intent to change it?

**Review by conradoplg — APPROVED** [review]
Looks good, I unresolved one conversation since I think you might have missing one change that you wanted to do, but I don't think it's required, so feel free to address it or mark as resolved

File: zebra-chain/src/parameters/network/tests/vectors.rs (hunk 20 above)
Comment (arya2) [reply zebra-chain/src/parameters/network/tests/vectors.rs:405]: Ah, it wasn't my intent to change it, but it should've been. We could've used `#[serde(default)]` there instead to allow the missing field in the configuration. It doesn't feel important enough to block merging the PR, but I'll update it in a follow-up PR for cleanups and addressing TODOs - unless there's an issue with CI that blocks this from merging immediately anyway.
